### PR TITLE
Viewport:  Add breakpoint variables and use them

### PR DESCRIPTION
**Pull request by geekman-rohit**
Adds Breakpoints variables and modifies typography to use them.

### CODE REVIEW

File: scss/atoms/typography/_typography.scss [hunk 1]
```diff
@@ -12,10 +12,10 @@
 body {
 	font: $normal #{ map-get( $defaultFont, font-size ) }/1.5 $sans;
 
-	@media ( max-width: 800px ) {
+	@media ( max-width: $viewport-sm ) {
```
Comment (robbbz) [inline scss/atoms/typography/_typography.scss]: I don't really like the `sm` abbreviation as new contributors are never going to know what this is.

Comment (geekman-rohit) [inline scss/atoms/typography/_typography.scss]: @robbbz actually it has become quite common in breakpoints and viewpoint related classes/terms. Most css frameworks use -md -sm -xs -lg for grid too. So I guess it should be fine.
But I am open to opinions, whatever most people on the team agree to.

File: scss/atoms/typography/_typography.scss [hunk 2]
```diff
@@ -12,10 +12,10 @@
...
 		font-size: 16px;
 	}
-	@media ( max-width: 500px ) {
+	@media ( max-width: $viewport-xs ) {
```
Comment (cvrebert) [inline scss/atoms/typography/_typography.scss]: Shouldn't these use `min-width`? Otherwise, won't the default on line 13 be dead code?

Comment (geekman-rohit) [inline scss/atoms/typography/_typography.scss]: @cvrebert yes it should be min, but it's kinda desktop first at the moment, with the default size applicable for anything larger than 720px screen width. We just decided in a meeting to go mobile first, I just thought that should be separate PR. But since both of them would be so tiny I should probably just push to this.

Comment (sfrisk) [inline scss/atoms/typography/_typography.scss]: Since this PR is fixing breakpoints to be variables anyway, I would just set this to be mobile first as well while we're in there.  Otherwise you're just repeating work with two PRs.

Comment (geekman-rohit) [inline scss/atoms/typography/_typography.scss]: Agreed.

**Comment by geekman-rohit** [comment]
@sfrisk I pushed to the PR.

File: scss/atoms/typography/_typography.scss [hunk 3]
```diff
@@ -12,11 +12,11 @@
 body {
 	font: $normal #{ map-get( $defaultFont, font-size ) }/1.5 $sans;
 
-	@media ( max-width: 800px ) {
+	@media ( min-width: $viewport-xs ) {
```
Comment (cvrebert) [inline scss/atoms/typography/_typography.scss]: Erm, I'd think that XS would start at 0, rather than having an implicit unnamed viewport size below XS.

File: scss/variables/breakpoints.js [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+	} else if ( typeof exports === "object" ) {
+		module.exports = factory( require( "./chassis" ) );
+	} else {
+		root.chassis = factory( root.chassis );
+	}
+}( this, function( chassis ) {
+
+chassis.breakpoints = {
+
+	"viewport-xs": {
```
Comment (cvrebert) [inline scss/variables/breakpoints.js]: I suggest using a `-min` or `-max` suffix for these variable names to clarify whether they're the start or the end of a viewport size tier.

File: scss/atoms/typography/_typography.scss (hunk 3 above)
Comment (geekman-rohit) [inline scss/atoms/typography/_typography.scss]: @cvrebert  umm the default starts at 14 px. Doesn't anything inside `@media ( min-width: $viewport-xs ) {` apply to screen sizes greater than $viewport-xs ?
 These variables are just breakpoints. with $viewport-sx defaulting to 400px.

Comment (geekman-rohit) [inline scss/atoms/typography/_typography.scss]: So it goes:
1. default:14px;
2. wider than 400px: 16px
3. wider than 720px: 20px

File: scss/variables/breakpoints.js (hunk 4 above)
Comment (geekman-rohit) [inline scss/variables/breakpoints.js]: Exactly why I wanted to call them $breakpoints-sm. We use these as breakpoints for viewports instead of specified min and max.

Comment (sfrisk) [inline scss/variables/breakpoints.js]: breakpoints make sense to me. If that will minimize confusion, I see no reason why we can't use that instead of 'viewports'

**Comment by geekman-rohit** [comment]
@sfrisk , @arschmitz 
I suggest we rename to $breakpoint-sm instead of $viewport as these are breakpoints.

File: scss/variables/breakpoints.js (hunk 4 above)
Comment (geekman-rohit) [inline scss/variables/breakpoints.js]: well we agreed on viewport in the meeting, Im open to anything everyone agrees to, though I vote for $breakpoint

File: scss/atoms/typography/_typography.scss (hunk 3 above)
Comment (cvrebert) [inline scss/atoms/typography/_typography.scss]: If we're explaining the breakpoints to a user of Chassis, right now it's:
* ???: 0px–399px
* XS: 400px–767px
* SM: 768px–919px
* etc...

I'm saying that having no name or reification of the "???" viewport size tier is awkward. I think it'd be better to conceptualize the "???" as XS and rename the other tiers accordingly.

(Sidenote: The "???" tier needs no media query since its minimum viewport width is 0px.)

File: scss/variables/breakpoints.js (hunk 4 above)
Comment (sfrisk) [inline scss/variables/breakpoints.js]: If it minimizes confusion among the larger community, go with `$breakpoint`.

Comment (arschmitz) [inline scss/variables/breakpoints.js]: i'm fine with that too

File: scss/atoms/typography/_typography.scss (hunk 3 above)
Comment (geekman-rohit) [inline scss/atoms/typography/_typography.scss]: Okay we currently do not explicitly define any viewports, we just use these breakpoints to apply styles to them cascadingly in mobile first manner, so the default font size in line 13 would apply to `???`
so to us its basically 
* less than breakpoint-xs (lets call them breakpoints now to avoid confusion)(This would imply 0-399) the font size in line 13 would apply to this
* more than or equal to breakpoint-sm 
* more than ...  breakpoint-md
* anything more ... than breakpoint lg
Right now this is the way we are distributing our viewport sizes

Comment (geekman-rohit) [inline scss/atoms/typography/_typography.scss]: If this approach could cause confusion to users/ contributors we could discuss this and change before merging

Comment (sfrisk) [inline scss/atoms/typography/_typography.scss]: I think it might make more sense that the min-width for this is the `$breakpoint-sm`, with the larger breakpoint for the font size 20px being the `$breakpoint-md`.  Sizing for this is a bit typography media query was done before we had formally decided on breakpoints, or having our media queries set up with "Mobile First" in mind.

As for the ???, I think the best way to maybe describe that might be "Default"  - it's at the $breakpoint-sm range where we we might want to start over-riding default styling with a media query.

Comment (geekman-rohit) [inline scss/atoms/typography/_typography.scss]: @sfrisk , I merely took the font sizes @MichaelArestad had applied to different screen sizes and did them mobile first.
So you mean we should have : 
1. default:14px;
2. wider than 768px: 16px
3. wider than 920px: 20px
??
that would meen a font size of 14px for the range 400-768 and 20 for 720-968

File: scss/atoms/typography/_typography.scss [hunk 5]
```diff
@@ -12,11 +12,11 @@
...
 		font-size: 16px;
 	}
-	@media ( max-width: 500px ) {
-		font-size: 14px;
+	@media ( min-width: $viewport-sm ) {
+		font-size: 20px;
```
Comment (cvrebert) [inline scss/atoms/typography/_typography.scss]: Shouldn't this be a variable?

File: scss/atoms/typography/_typography.scss (hunk 3 above)
Comment (sfrisk) [inline scss/atoms/typography/_typography.scss]: Yes that is what I am suggesting., @MichaelArestad's breakpoints were at amounts higher than what we eventually agreed on breakpoints, and 400 seems a bit small to me for a breakpoint for the smallest font size.

Comment (geekman-rohit) [inline scss/atoms/typography/_typography.scss]: then we should have the smallest font size to 16 px too, 14 is kind of too small I think

File: scss/atoms/typography/_typography.scss (hunk 5 above)
Comment (sfrisk) [inline scss/atoms/typography/_typography.scss]: Yup. Typography was created before jsass became a thing we were using, we need to go through Typography and change a few things to variables.

File: scss/atoms/typography/_typography.scss (hunk 3 above)
Comment (sfrisk) [inline scss/atoms/typography/_typography.scss]: Could remove one of the media queries then. Also like @cvrebert suggested below, make it a variable.

Comment (geekman-rohit) [inline scss/atoms/typography/_typography.scss]: We should change them all to variables and have media queries for all viewports so we could allow user to override in the themeroller

Comment (arschmitz) [inline scss/atoms/typography/_typography.scss]: I agree with @cvrebert on this about that the smallest one ??? in his example should be `xs`. My suggestion for naming was based on that. I like and find the breakpoints from flexbox grid very user friendly. see https://github.com/kristoferjoseph/flexboxgrid/blob/master/src/css/flexboxgrid.css
notice `xs` has not explicit size definition also as @cvrebert suggests.

Comment (arschmitz) [inline scss/atoms/typography/_typography.scss]: Just mentioning this again though from way back. I still think its a bad idea to adjust the default font size in a media query. I think this is too heavy handed to apply in general. Seems like something that should be project specific if its going to be done.

Comment (geekman-rohit) [inline scss/atoms/typography/_typography.scss]: @arschmitz well now I understand why you were insisting on viewport, looking at it from that angle we should have "viewport-xs-min & viewport-xs-max" and so on, a total of 8 variables. I just thought we could achieve the same with only 4 breakpoint variables used to separate viewports insteak of explixitly stating their boundries. Since most people seem to agree on the 'viewport' approach we should adopt that

Comment (geekman-rohit) [inline scss/atoms/typography/_typography.scss]: about font sizes, I donot know about disadvantages of changing font size in media queries, but I  dothink using a media query there is unnecessary. A font size of 16/17px should be good for all screens, even 1080p displays.

**Comment by geekman-rohit** [comment]
viewport-xs-min : 0
viewport-sm-min : 400
viewport-md-min: 768
viewport-lg-min: 920
viewport-lg-max: 1200

if everyone approves to that I could make the change and update ?

**Comment by geekman-rohit** [comment]
@cvrebert is that (my earlier comment ) what you were suggesting?

**Comment by cvrebert** [comment]
@geekman-rohit Yeah, pretty much. Although the minimums being odd numbers seems slightly weird.

**Comment by geekman-rohit** [comment]
@sfrisk , my point exactly. min should start at 400, xs ending at 399.

**Comment by arschmitz** [comment]
@cvrebert i agree

**Comment by sfrisk** [comment]
Then do that.

**Comment by geekman-rohit** [comment]
@sfrisk any reviews on this?

**Comment by sfrisk** [comment]
@geekman-rohit - remove the changes on files that aren't on breakpoint.js and make sure it works with the latest version of master, since I'm pretty sure this branch is a bit out of date.  Once that is done, I can pull in these changes.

File: scss/atoms/typography/_typography.scss [hunk 6]
```diff
@@ -12,10 +12,10 @@
 body {
 	font: $normal #{ map-get( $defaultFont, font-size ) }/1.5 $sans;
 
-	@media ( max-width: 800px ) {
+	@media ( max-width: $viewport-md-min ) {
 		font-size: 16px;
```
Comment (geekman-rohit) [inline scss/atoms/typography/_typography.scss]: Not mobile first currently, We can have that done in another PR. I simply used the variables instead of direct pixel values.

**Comment by geekman-rohit** [comment]
@sfrisk I think this can go in now?

**Comment by sfrisk** [comment]
Bahhhhh, I missed a step when I merged this in. My bad guys.
